javatests/arcs/core/storage/ReferenceModeStoreDatabaseIntegrationTest.kt

```diff
@@ -0,0 +1,546 @@
+            SchemaFields(
+                singletons = mapOf("name" to FieldType.Text, "age" to FieldType.Number),
+                collections = emptyMap()
+            ),
+            SchemaDescription(),
+            hash
+        )
+
+        testKey = ReferenceModeStorageKey(
+            DatabaseStorageKey("entities", hash),
+            DatabaseStorageKey("set", hash)
+        )
+
+        DriverFactory.clearRegistrationsForTesting()
+        databaseFactory = MockDatabaseFactory()
+        DatabaseDriverProvider.configure(databaseFactory) { schema }
+    }
+
+    @Test
+    fun propagatesModelUpdates_fromProxies_toDrivers() = runBlockingTest {
+        val activeStore = createReferenceModeStore()
+
+        val collection = CrdtSet<RawEntity>()
+        val entity = createPersonEntity("an-id", "bob", 42)
+        collection.applyOperation(
+            CrdtSet.Operation.Add("me", VersionMap("me" to 1), entity)
+        )
+
+        logRule("Sending ModelUpdate")
+
+        assertThat(
+            activeStore.onProxyMessage(
+                ProxyMessage.ModelUpdate(RefModeStoreData.Set(collection.data), 1)
+            )
+        ).isTrue()
+
+        logRule("ModelUpdate sent")
+
+        val actor = activeStore.crdtKey
+        val containerKey = activeStore.containerStore.storageKey as DatabaseStorageKey
+        val database = databaseFactory.getDatabase(containerKey.dbName, containerKey.persistent)
+
+        val capturedCollection = requireNotNull(
+            database.get(containerKey, DatabaseData.Collection::class)
+        ) as DatabaseData.Collection
+
+        assertThat(capturedCollection.values)
+            .containsExactly(
+                Reference("an-id", activeStore.backingStore.storageKey, VersionMap(actor to 1))
+            )
+
+        val bobKey = activeStore.backingStore.storageKey.childKeyWithComponent("an-id")
+        val capturedBob = requireNotNull(
+            database.get(bobKey, DatabaseData.Entity::class) as? DatabaseData.Entity
+        )
+
+        assertThat(capturedBob.entity.data["name"]).isEqualTo("bob")
+        assertThat(capturedBob.entity.data["age"]).isEqualTo(42)
+    }
+
+    @Test
+    fun canCloneData_fromAnotherStore() = runBlockingTest {
+        val activeStore = createReferenceModeStore()
+
+        // Add some data.
+        val collection = CrdtSet<RawEntity>()
+        val entity = createPersonEntity("an-id", "bob", 42)
+        collection.applyOperation(
+            CrdtSet.Operation.Add("me", VersionMap("me" to 1), entity)
+        )
+        activeStore.onProxyMessage(
+            ProxyMessage.ModelUpdate(RefModeStoreData.Set(collection.data), 1)
+        )
+
+        // Clone
+        val activeStore2 = createReferenceModeStore()
+        activeStore2.cloneFrom(activeStore)
+
+        assertThat(activeStore2.getLocalData()).isEqualTo(activeStore.getLocalData())
+        assertThat(activeStore2.getLocalData() === activeStore.getLocalData()).isFalse()
```

Comment:
isNotSameInstanceAs

Reply:
Ah that's what I get for copy-pasting the bulk of this 🙃 I'll fix it in the main RefModeStoreIntegrationTest as well.